schema.c

```diff
@@ -4065,3 +4381,517 @@ sch_json_to_gnode (sch_instance * instance, sch_node * schema, json_t * json, in
+                             int flags, bool exact, GNode **node, sch_xlat_data *x_data,
+                             sch_node **rschema)
+{
+    GNode *nnode;
+    GNode *int_node;
+    GNode *ret_node;
+    GNode *remaining = NULL;
+    GNode *last_match_node = NULL;
+    bool match = false;
+
+    if (xlat_tree_match (start_node, x_data->ext_tree, &last_match_node))
+    {
+        if (g_hash_table_lookup (node_table, last_match_node))
+            return match;
+
+        sch_translate_mark_node (node_table, last_match_node);
+        if (last_match_node && last_match_node->children)
+        {
+            remaining = g_node_copy_deep (last_match_node->children, copy_node_data, NULL);
+            sch_translate_mark_children (node_table, last_match_node);
+        }
+
+        int_node = x_data->int_tree;
+        if (rschema && !*rschema)
+        {
+            *rschema = sch_traverse_get_schema (instance, int_node, flags);
+        }
+
+        if (!*node)
+            *node = g_node_new (g_strdup (APTERYX_NAME (int_node)));
+        nnode = *node;
+        int_node = int_node->children;
+        while (int_node)
+        {
+            if (g_strcmp0 (APTERYX_NAME (int_node), "*") == 0)
+            {
+                char *str = sch_translate_get_wildcard_data (start_node, x_data->ext_tree, 1);
+                if (sch_translate_child_exists (nnode, str, &ret_node))
+                {
+                    g_free (str);
+                    nnode = ret_node;
+                }
+                else
+                    nnode = g_node_append_data (nnode, str);
+            }
+            else
+            {
+                if (sch_translate_child_exists (nnode, APTERYX_NAME (int_node), &ret_node))
+                    nnode = ret_node;
+                else
+                    nnode = g_node_append_data (nnode, g_strdup (APTERYX_NAME (int_node)));
+            }
+
+            int_node = int_node->children;
+        }
+        if (remaining)
+        {
+            if (sch_translate_child_exists (nnode, APTERYX_NAME (remaining), &ret_node))
+            {
+                nnode = g_node_append (ret_node, remaining);
+            }
+            else
+                nnode = g_node_append (nnode, remaining);
+        }
+
+        if (x_data->lua_module && nnode)
+            sch_translate_data (instance, nnode->parent, x_data, flags, true);
+
+        match = true;
+    }
+
+    return match;
+}
+
+GNode *
+sch_translate_input (sch_instance * instance, GNode *node, int flags, void **xlat_data, sch_node **rschema)
+{
+    GNode *ret_node = NULL;
+    char *name;
+    sch_xlat_root *root;
+    GHashTable *node_table = NULL;
+    GList *list;
+    sch_xlat_data *x_data;
+    bool match;
+    int num_nodes = 0;
+    int query_depth = 0;
+
+    *xlat_data = NULL;
+    if (!instance->xlat_hash_table || !node)
+        return node;
+
+    name = APTERYX_NAME (node);
```

Comment:
what happens if the root node has a path in it /node1/node2/node3

Reply:
At the moment roots are based only on the name of the first node in the path. So from your example "node1". This seemed to be enough for our various interface models.